### PR TITLE
Keystore

### PR DESCRIPTION
**Pull request by yosriz**
* add secure keystore implementation
* use Android KeyStore with RSA for Android 18-22, AES for 23+,
 and non-secure for 16-17
* write tests for the different KeyStore implementations, tests should
 run on the proper platform, o.w. tests will be skipped

### CODE REVIEW

File: kin-sdk-core/src/test/java/kin/sdk/core/KeyStoreImplTest.java [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+            throw new CryptoException(new Exception("Some bad things happened"));
+        }
+
+        @Override
+        public String decrypt(String encryptedSecret) throws CryptoException {
+            throw new CryptoException(new Exception("Some bad things happened"));
+        }
+    }
+
+    private static class ExceptionThrowerStore implements Store {
```
Comment (berryve) [inline kin-sdk-core/src/test/java/kin/sdk/core/KeyStoreImplTest.java:63]: IllegalJsonStore might be a better name

File: kin-sdk-core/src/main/java/kin/sdk/core/KinAccount.java [hunk 2]
```diff
@@ -15,13 +15,11 @@
...
-     * Exports the keystore json file
+     * Deprecated, export will return null
      *
-     * @param passphrase the passphrase used to create the account
-     * @param newPassphrase the exported json will be encrypted using this new passphrase. The original keystore and
-     * passphrase will not change.
-     * @return String the json string
+     * @return null
      */
+    @Deprecated
```
Comment (berryve) [inline kin-sdk-core/src/main/java/kin/sdk/core/KinAccount.java]: let's delete this completely

File: kin-sdk-core/src/main/java/kin/sdk/core/ClientWrapper.java [hunk 3]
```diff
@@ -17,6 +17,7 @@
  */
 class ClientWrapper {
```
Comment (berryve) [inline kin-sdk-core/src/main/java/kin/sdk/core/ClientWrapper.java:19]: remove ClientWrapper ?

File: kin-sdk-core/src/main/java/kin/sdk/core/EncryptorFactory.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
+package kin.sdk.core;
+
+import android.content.Context;
+import android.os.Build;
+import android.os.Build.VERSION_CODES;
+
+class EncryptorFactory {
+
+    private static final String VERSION_KEY = "encryptor_ver";
+    private static final String VER_23 = "23";
```
Comment (berryve) [inline kin-sdk-core/src/main/java/kin/sdk/core/EncryptorFactory.java]: can you use android constants?

File: kin-sdk-core/src/main/java/kin/sdk/core/EncryptorFactory.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+    private static final String VER_18 = "18";
+    private static final String VER_16 = "16";
+
+    //no instances
+    private EncryptorFactory() {
+    }
+
+    static Encryptor create(Context context, Store store) {
+        String versionString = store.getString(VERSION_KEY);
```
Comment (berryve) [inline kin-sdk-core/src/main/java/kin/sdk/core/EncryptorFactory.java]: you can simplify the logic here by having a separate method that returns the version to be used.

Then you can simply switch over the version to decide which encryptor to use

File: kin-sdk-core/src/main/java/kin/sdk/core/EncryptorImplV18.java [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+    private final Context context;
+
+    EncryptorImplV18(Context context) {
+        this.context = context;
+    }
+
+    public String encrypt(String secret) throws CryptoException {
+        try {
+            KeyStore keyStore = KeyStore.getInstance(ANDROID_KEY_STORE);
+            keyStore.load(null);
```
Comment (berryve) [inline kin-sdk-core/src/main/java/kin/sdk/core/EncryptorImplV18.java:48]: ?

File: kin-sdk-core/src/main/java/kin/sdk/core/EncryptorImplV16.java [hunk 7]
```diff
@@ -0,0 +1,18 @@
+package kin.sdk.core;
+
+
+/**
+ * UNSAFE no-op encryptor for below api 18, it will do no encryption
+ */
+class EncryptorImplV16 implements Encryptor {
```
Comment (berryve) [inline kin-sdk-core/src/main/java/kin/sdk/core/EncryptorImplV16.java:7]: add disclaimer on readme for this

File: kin-sdk-core/src/main/java/kin/sdk/core/EncryptorFactory.java (hunk 5 above)
Comment (berryve) [reply kin-sdk-core/src/main/java/kin/sdk/core/EncryptorFactory.java]: no option to upgrade encryptor - let product know

File: kin-sdk-core/src/main/java/kin/sdk/core/EncryptorImplV18.java [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+        byte[] encryptedBytes = performEncryption(secret.getBytes("UTF-8"), publicKey);
+        return Base64.encodeToString(encryptedBytes, Base64.DEFAULT);
+    }
+
+    @TargetApi(VERSION_CODES.JELLY_BEAN_MR2)
+    private KeyPair generateRsaPair()
+        throws NoSuchAlgorithmException, NoSuchProviderException, InvalidAlgorithmParameterException {
+        Calendar start = Calendar.getInstance();
+        Calendar end = Calendar.getInstance();
+        end.add(Calendar.YEAR, 30);
```
Comment (berryve) [inline kin-sdk-core/src/main/java/kin/sdk/core/EncryptorImplV18.java]: bug 30? let product know

File: kin-sdk-core/src/androidTest/java/kin/sdk/core/EncryptorImplV23Test.java [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+import static junit.framework.Assert.assertEquals;
+import static org.junit.Assert.assertNotEquals;
+import static org.junit.Assume.assumeTrue;
+
+import android.os.Build;
+import android.os.Build.VERSION_CODES;
+import org.junit.Before;
+import org.junit.Test;
+
+public class EncryptorImplV23Test {
```
Comment (berryve) [inline kin-sdk-core/src/androidTest/java/kin/sdk/core/EncryptorImplV23Test.java:12]: check if you can change emulator android version that runs tests

**Review by berryve — APPROVED** [review]
fix comments...

File: kin-sdk-core/src/main/java/kin/sdk/core/EncryptorImplV18.java (hunk 6 above)
Comment (yosriz) [reply kin-sdk-core/src/main/java/kin/sdk/core/EncryptorImplV18.java:48]: that's a load method with load params that can be null, in our case, it should be null

File: kin-sdk-core/src/main/java/kin/sdk/core/EncryptorImplV16.java (hunk 7 above)
Comment (yosriz) [reply kin-sdk-core/src/main/java/kin/sdk/core/EncryptorImplV16.java:7]: I will took it as a separate discussion how to handle version 16

File: kin-sdk-core/src/androidTest/java/kin/sdk/core/EncryptorImplV23Test.java (hunk 9 above)
Comment (yosriz) [reply kin-sdk-core/src/androidTest/java/kin/sdk/core/EncryptorImplV23Test.java:12]: I have additional problem with android tests, will investigate it in another task

**Pull request merged**